// @tiptap/extension-underline/dist/tiptap-extension-underline.esm.js
import { Mark, mergeAttributes } from '@tiptap/core';

const Underline = Mark.create({
    name: 'underline',
    defaultOptions: {
        HTMLAttributes: {},
    },
    parseHTML() {
        return [
            {
                tag: 'u',
            },
            {
                style: 'text-decoration',
                consuming: false,
                getAttrs: style => (style.includes('underline') ? {} : false),
            },
        ];
    },
    renderHTML({ HTMLAttributes }) {
        return ['u', mergeAttributes(this.options.HTMLAttributes, HTMLAttributes), 0];
    },
    addCommands() {
        return {
            setUnderline: () => ({ commands }) => {
                return commands.setMark('underline');
            },
            toggleUnderline: () => ({ commands }) => {
                return commands.toggleMark('underline');
            },
            unsetUnderline: () => ({ commands }) => {
                return commands.unsetMark('underline');
            },
        };
    },
    addKeyboardShortcuts() {
        return {
            'Mod-u': () => this.editor.commands.toggleUnderline(),
        };
    },
});

export default Underline;
export { Underline };
