// @tiptap/extension-underline/dist/tiptap-extension-underline.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@tiptap/core');

const Underline = core.Mark.create({
    name: 'underline',
    defaultOptions: {
        HTMLAttributes: {},
    },
    parseHTML() {
        return [
            {
                tag: 'u',
            },
            {
                style: 'text-decoration',
                consuming: false,
                getAttrs: style => (style.includes('underline') ? {} : false),
            },
        ];
    },
    renderHTML({ HTMLAttributes }) {
        return ['u', core.mergeAttributes(this.options.HTMLAttributes, HTMLAttributes), 0];
    },
    addCommands() {
        return {
            setUnderline: () => ({ commands }) => {
                return commands.setMark('underline');
            },
            toggleUnderline: () => ({ commands }) => {
                return commands.toggleMark('underline');
            },
            unsetUnderline: () => ({ commands }) => {
                return commands.unsetMark('underline');
            },
        };
    },
    addKeyboardShortcuts() {
        return {
            'Mod-u': () => this.editor.commands.toggleUnderline(),
        };
    },
});

exports.Underline = Underline;
exports.default = Underline;
