// @tiptap/extension-italic/dist/tiptap-extension-italic.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@tiptap/core');

const starInputRegex = /(?:^|\s)((?:\*)((?:[^*]+))(?:\*))$/gm;
const starPasteRegex = /(?:^|\s)((?:\*)((?:[^*]+))(?:\*))/gm;
const underscoreInputRegex = /(?:^|\s)((?:_)((?:[^_]+))(?:_))$/gm;
const underscorePasteRegex = /(?:^|\s)((?:_)((?:[^_]+))(?:_))/gm;
const Italic = core.Mark.create({
    name: 'italic',
    defaultOptions: {
        HTMLAttributes: {},
    },
    parseHTML() {
        return [
            {
                tag: 'em',
            },
            {
                tag: 'i',
                getAttrs: node => node.style.fontStyle !== 'normal' && null,
            },
            {
                style: 'font-style=italic',
            },
        ];
    },
    renderHTML({ HTMLAttributes }) {
        return ['em', core.mergeAttributes(this.options.HTMLAttributes, HTMLAttributes), 0];
    },
    addCommands() {
        return {
            setItalic: () => ({ commands }) => {
                return commands.setMark('italic');
            },
            toggleItalic: () => ({ commands }) => {
                return commands.toggleMark('italic');
            },
            unsetItalic: () => ({ commands }) => {
                return commands.unsetMark('italic');
            },
        };
    },
    addKeyboardShortcuts() {
        return {
            'Mod-i': () => this.editor.commands.toggleItalic(),
        };
    },
    addInputRules() {
        return [
            core.markInputRule(starInputRegex, this.type),
            core.markInputRule(underscoreInputRegex, this.type),
        ];
    },
    addPasteRules() {
        return [
            core.markPasteRule(starPasteRegex, this.type),
            core.markPasteRule(underscorePasteRegex, this.type),
        ];
    },
});

exports.Italic = Italic;
exports.default = Italic;
exports.starInputRegex = starInputRegex;
exports.starPasteRegex = starPasteRegex;
exports.underscoreInputRegex = underscoreInputRegex;
exports.underscorePasteRegex = underscorePasteRegex;
