// @tiptap/extension-italic/dist/tiptap-extension-italic.umd.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tiptap/core')) :
  typeof define === 'function' && define.amd ? define(['exports', '@tiptap/core'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global['@tiptap/extension-italic'] = {}, global.core));
}(this, (function (exports, core) { 'use strict';

  const starInputRegex = /(?:^|\s)((?:\*)((?:[^*]+))(?:\*))$/gm;
  const starPasteRegex = /(?:^|\s)((?:\*)((?:[^*]+))(?:\*))/gm;
  const underscoreInputRegex = /(?:^|\s)((?:_)((?:[^_]+))(?:_))$/gm;
  const underscorePasteRegex = /(?:^|\s)((?:_)((?:[^_]+))(?:_))/gm;
  const Italic = core.Mark.create({
      name: 'italic',
      defaultOptions: {
          HTMLAttributes: {},
      },
      parseHTML() {
          return [
              {
                  tag: 'em',
              },
              {
                  tag: 'i',
                  getAttrs: node => node.style.fontStyle !== 'normal' && null,
              },
              {
                  style: 'font-style=italic',
              },
          ];
      },
      renderHTML({ HTMLAttributes }) {
          return ['em', core.mergeAttributes(this.options.HTMLAttributes, HTMLAttributes), 0];
      },
      addCommands() {
          return {
              setItalic: () => ({ commands }) => {
                  return commands.setMark('italic');
              },
              toggleItalic: () => ({ commands }) => {
                  return commands.toggleMark('italic');
              },
              unsetItalic: () => ({ commands }) => {
                  return commands.unsetMark('italic');
              },
          };
      },
      addKeyboardShortcuts() {
          return {
              'Mod-i': () => this.editor.commands.toggleItalic(),
          };
      },
      addInputRules() {
          return [
              core.markInputRule(starInputRegex, this.type),
              core.markInputRule(underscoreInputRegex, this.type),
          ];
      },
      addPasteRules() {
          return [
              core.markPasteRule(starPasteRegex, this.type),
              core.markPasteRule(underscorePasteRegex, this.type),
          ];
      },
  });

  exports.Italic = Italic;
  exports.default = Italic;
  exports.starInputRegex = starInputRegex;
  exports.starPasteRegex = starPasteRegex;
  exports.underscoreInputRegex = underscoreInputRegex;
  exports.underscorePasteRegex = underscorePasteRegex;

  Object.defineProperty(exports, '__esModule', { value: true });

})));
