// @tiptap/extension-list-item/dist/tiptap-extension-list-item.umd.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tiptap/core')) :
  typeof define === 'function' && define.amd ? define(['exports', '@tiptap/core'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global['@tiptap/extension-list-item'] = {}, global.core));
}(this, (function (exports, core) { 'use strict';

  const ListItem = core.Node.create({
      name: 'listItem',
      defaultOptions: {
          HTMLAttributes: {},
      },
      content: 'paragraph block*',
      defining: true,
      parseHTML() {
          return [
              {
                  tag: 'li',
              },
          ];
      },
      renderHTML({ HTMLAttributes }) {
          return ['li', core.mergeAttributes(this.options.HTMLAttributes, HTMLAttributes), 0];
      },
      addKeyboardShortcuts() {
          return {
              Enter: () => this.editor.commands.splitListItem('listItem'),
              Tab: () => this.editor.commands.sinkListItem('listItem'),
              'Shift-Tab': () => this.editor.commands.liftListItem('listItem'),
          };
      },
  });

  exports.ListItem = ListItem;
  exports.default = ListItem;

  Object.defineProperty(exports, '__esModule', { value: true });

})));
