// @tiptap/extension-strike/dist/tiptap-extension-strike.esm.js
import { Mark, mergeAttributes, markInputRule, markPasteRule } from '@tiptap/core';

const inputRegex = /(?:^|\s)((?:~~)((?:[^~]+))(?:~~))$/gm;
const pasteRegex = /(?:^|\s)((?:~~)((?:[^~]+))(?:~~))/gm;
const Strike = Mark.create({
    name: 'strike',
    defaultOptions: {
        HTMLAttributes: {},
    },
    parseHTML() {
        return [
            {
                tag: 's',
            },
            {
                tag: 'del',
            },
            {
                tag: 'strike',
            },
            {
                style: 'text-decoration',
                consuming: false,
                getAttrs: style => (style.includes('line-through') ? {} : false),
            },
        ];
    },
    renderHTML({ HTMLAttributes }) {
        return ['s', mergeAttributes(this.options.HTMLAttributes, HTMLAttributes), 0];
    },
    addCommands() {
        return {
            setStrike: () => ({ commands }) => {
                return commands.setMark('strike');
            },
            toggleStrike: () => ({ commands }) => {
                return commands.toggleMark('strike');
            },
            unsetStrike: () => ({ commands }) => {
                return commands.unsetMark('strike');
            },
        };
    },
    addKeyboardShortcuts() {
        return {
            'Mod-Shift-x': () => this.editor.commands.toggleStrike(),
        };
    },
    addInputRules() {
        return [
            markInputRule(inputRegex, this.type),
        ];
    },
    addPasteRules() {
        return [
            markPasteRule(pasteRegex, this.type),
        ];
    },
});

export default Strike;
export { Strike, inputRegex, pasteRegex };
