// @tiptap/extension-strike/dist/tiptap-extension-strike.umd.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tiptap/core')) :
  typeof define === 'function' && define.amd ? define(['exports', '@tiptap/core'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global['@tiptap/extension-strike'] = {}, global.core));
}(this, (function (exports, core) { 'use strict';

  const inputRegex = /(?:^|\s)((?:~~)((?:[^~]+))(?:~~))$/gm;
  const pasteRegex = /(?:^|\s)((?:~~)((?:[^~]+))(?:~~))/gm;
  const Strike = core.Mark.create({
      name: 'strike',
      defaultOptions: {
          HTMLAttributes: {},
      },
      parseHTML() {
          return [
              {
                  tag: 's',
              },
              {
                  tag: 'del',
              },
              {
                  tag: 'strike',
              },
              {
                  style: 'text-decoration',
                  consuming: false,
                  getAttrs: style => (style.includes('line-through') ? {} : false),
              },
          ];
      },
      renderHTML({ HTMLAttributes }) {
          return ['s', core.mergeAttributes(this.options.HTMLAttributes, HTMLAttributes), 0];
      },
      addCommands() {
          return {
              setStrike: () => ({ commands }) => {
                  return commands.setMark('strike');
              },
              toggleStrike: () => ({ commands }) => {
                  return commands.toggleMark('strike');
              },
              unsetStrike: () => ({ commands }) => {
                  return commands.unsetMark('strike');
              },
          };
      },
      addKeyboardShortcuts() {
          return {
              'Mod-Shift-x': () => this.editor.commands.toggleStrike(),
          };
      },
      addInputRules() {
          return [
              core.markInputRule(inputRegex, this.type),
          ];
      },
      addPasteRules() {
          return [
              core.markPasteRule(pasteRegex, this.type),
          ];
      },
  });

  exports.Strike = Strike;
  exports.default = Strike;
  exports.inputRegex = inputRegex;
  exports.pasteRegex = pasteRegex;

  Object.defineProperty(exports, '__esModule', { value: true });

})));
