// @tiptap/extension-bullet-list/dist/tiptap-extension-bullet-list.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@tiptap/core');
var prosemirrorInputrules = require('prosemirror-inputrules');

const inputRegex = /^\s*([-+*])\s$/;
const BulletList = core.Node.create({
    name: 'bulletList',
    defaultOptions: {
        HTMLAttributes: {},
    },
    group: 'block list',
    content: 'listItem+',
    parseHTML() {
        return [
            { tag: 'ul' },
        ];
    },
    renderHTML({ HTMLAttributes }) {
        return ['ul', core.mergeAttributes(this.options.HTMLAttributes, HTMLAttributes), 0];
    },
    addCommands() {
        return {
            toggleBulletList: () => ({ commands }) => {
                return commands.toggleList('bulletList', 'listItem');
            },
        };
    },
    addKeyboardShortcuts() {
        return {
            'Mod-Shift-8': () => this.editor.commands.toggleBulletList(),
        };
    },
    addInputRules() {
        return [
            prosemirrorInputrules.wrappingInputRule(inputRegex, this.type),
        ];
    },
});

exports.BulletList = BulletList;
exports.default = BulletList;
exports.inputRegex = inputRegex;
