// @tiptap/extension-ordered-list/dist/tiptap-extension-ordered-list.umd.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tiptap/core'), require('prosemirror-inputrules')) :
  typeof define === 'function' && define.amd ? define(['exports', '@tiptap/core', 'prosemirror-inputrules'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global['@tiptap/extension-ordered-list'] = {}, global.core, global.prosemirrorInputrules));
}(this, (function (exports, core, prosemirrorInputrules) { 'use strict';

  const inputRegex = /^(\d+)\.\s$/;
  const OrderedList = core.Node.create({
      name: 'orderedList',
      defaultOptions: {
          HTMLAttributes: {},
      },
      group: 'block list',
      content: 'listItem+',
      addAttributes() {
          return {
              start: {
                  default: 1,
                  parseHTML: element => {
                      return element.hasAttribute('start')
                          ? parseInt(element.getAttribute('start') || '', 10)
                          : 1;
                  },
              },
          };
      },
      parseHTML() {
          return [
              {
                  tag: 'ol',
              },
          ];
      },
      renderHTML({ HTMLAttributes }) {
          const { start, ...attributesWithoutStart } = HTMLAttributes;
          return start === 1
              ? ['ol', core.mergeAttributes(this.options.HTMLAttributes, attributesWithoutStart), 0]
              : ['ol', core.mergeAttributes(this.options.HTMLAttributes, HTMLAttributes), 0];
      },
      addCommands() {
          return {
              toggleOrderedList: () => ({ commands }) => {
                  return commands.toggleList('orderedList', 'listItem');
              },
          };
      },
      addKeyboardShortcuts() {
          return {
              'Mod-Shift-7': () => this.editor.commands.toggleOrderedList(),
          };
      },
      addInputRules() {
          return [
              prosemirrorInputrules.wrappingInputRule(inputRegex, this.type, match => ({ start: +match[1] }), (match, node) => node.childCount + node.attrs.start === +match[1]),
          ];
      },
  });

  exports.OrderedList = OrderedList;
  exports['default'] = OrderedList;
  exports.inputRegex = inputRegex;

  Object.defineProperty(exports, '__esModule', { value: true });

})));
