// @tiptap/extension-bold/dist/tiptap-extension-bold.esm.js
import { Mark, mergeAttributes, markInputRule, markPasteRule } from '@tiptap/core';

const starInputRegex = /(?:^|\s)((?:\*\*)((?:[^*]+))(?:\*\*))$/gm;
const starPasteRegex = /(?:^|\s)((?:\*\*)((?:[^*]+))(?:\*\*))/gm;
const underscoreInputRegex = /(?:^|\s)((?:__)((?:[^__]+))(?:__))$/gm;
const underscorePasteRegex = /(?:^|\s)((?:__)((?:[^__]+))(?:__))/gm;
const Bold = Mark.create({
    name: 'bold',
    defaultOptions: {
        HTMLAttributes: {},
    },
    parseHTML() {
        return [
            {
                tag: 'strong',
            },
            {
                tag: 'b',
                getAttrs: node => node.style.fontWeight !== 'normal' && null,
            },
            {
                style: 'font-weight',
                getAttrs: value => /^(bold(er)?|[5-9]\d{2,})$/.test(value) && null,
            },
        ];
    },
    renderHTML({ HTMLAttributes }) {
        return ['strong', mergeAttributes(this.options.HTMLAttributes, HTMLAttributes), 0];
    },
    addCommands() {
        return {
            setBold: () => ({ commands }) => {
                return commands.setMark('bold');
            },
            toggleBold: () => ({ commands }) => {
                return commands.toggleMark('bold');
            },
            unsetBold: () => ({ commands }) => {
                return commands.unsetMark('bold');
            },
        };
    },
    addKeyboardShortcuts() {
        return {
            'Mod-b': () => this.editor.commands.toggleBold(),
        };
    },
    addInputRules() {
        return [
            markInputRule(starInputRegex, this.type),
            markInputRule(underscoreInputRegex, this.type),
        ];
    },
    addPasteRules() {
        return [
            markPasteRule(starPasteRegex, this.type),
            markPasteRule(underscorePasteRegex, this.type),
        ];
    },
});

export default Bold;
export { Bold, starInputRegex, starPasteRegex, underscoreInputRegex, underscorePasteRegex };
