// @tiptap/extension-code/dist/tiptap-extension-code.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@tiptap/core');

const inputRegex = /(?:^|\s)((?:`)((?:[^`]+))(?:`))$/gm;
const pasteRegex = /(?:^|\s)((?:`)((?:[^`]+))(?:`))/gm;
const Code = core.Mark.create({
    name: 'code',
    defaultOptions: {
        HTMLAttributes: {},
    },
    excludes: '_',
    parseHTML() {
        return [
            { tag: 'code' },
        ];
    },
    renderHTML({ HTMLAttributes }) {
        return ['code', core.mergeAttributes(this.options.HTMLAttributes, HTMLAttributes), 0];
    },
    addCommands() {
        return {
            setCode: () => ({ commands }) => {
                return commands.setMark('code');
            },
            toggleCode: () => ({ commands }) => {
                return commands.toggleMark('code');
            },
            unsetCode: () => ({ commands }) => {
                return commands.unsetMark('code');
            },
        };
    },
    addKeyboardShortcuts() {
        return {
            'Mod-e': () => this.editor.commands.toggleCode(),
        };
    },
    addInputRules() {
        return [
            core.markInputRule(inputRegex, this.type),
        ];
    },
    addPasteRules() {
        return [
            core.markPasteRule(pasteRegex, this.type),
        ];
    },
});

exports.Code = Code;
exports['default'] = Code;
exports.inputRegex = inputRegex;
exports.pasteRegex = pasteRegex;
