// @tiptap/extension-code/dist/tiptap-extension-code.esm.js
import { Mark, mergeAttributes, markInputRule, markPasteRule } from '@tiptap/core';

const inputRegex = /(?:^|\s)((?:`)((?:[^`]+))(?:`))$/gm;
const pasteRegex = /(?:^|\s)((?:`)((?:[^`]+))(?:`))/gm;
const Code = Mark.create({
    name: 'code',
    defaultOptions: {
        HTMLAttributes: {},
    },
    excludes: '_',
    parseHTML() {
        return [
            { tag: 'code' },
        ];
    },
    renderHTML({ HTMLAttributes }) {
        return ['code', mergeAttributes(this.options.HTMLAttributes, HTMLAttributes), 0];
    },
    addCommands() {
        return {
            setCode: () => ({ commands }) => {
                return commands.setMark('code');
            },
            toggleCode: () => ({ commands }) => {
                return commands.toggleMark('code');
            },
            unsetCode: () => ({ commands }) => {
                return commands.unsetMark('code');
            },
        };
    },
    addKeyboardShortcuts() {
        return {
            'Mod-e': () => this.editor.commands.toggleCode(),
        };
    },
    addInputRules() {
        return [
            markInputRule(inputRegex, this.type),
        ];
    },
    addPasteRules() {
        return [
            markPasteRule(pasteRegex, this.type),
        ];
    },
});

export { Code, Code as default, inputRegex, pasteRegex };
