// @tiptap/extension-blockquote/dist/tiptap-extension-blockquote.umd.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tiptap/core'), require('prosemirror-inputrules')) :
  typeof define === 'function' && define.amd ? define(['exports', '@tiptap/core', 'prosemirror-inputrules'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global['@tiptap/extension-blockquote'] = {}, global.core, global.prosemirrorInputrules));
}(this, (function (exports, core, prosemirrorInputrules) { 'use strict';

  const inputRegex = /^\s*>\s$/gm;
  const Blockquote = core.Node.create({
      name: 'blockquote',
      defaultOptions: {
          HTMLAttributes: {},
      },
      content: 'block*',
      group: 'block',
      defining: true,
      parseHTML() {
          return [
              { tag: 'blockquote' },
          ];
      },
      renderHTML({ HTMLAttributes }) {
          return ['blockquote', core.mergeAttributes(this.options.HTMLAttributes, HTMLAttributes), 0];
      },
      addCommands() {
          return {
              setBlockquote: () => ({ commands }) => {
                  return commands.wrapIn('blockquote');
              },
              toggleBlockquote: () => ({ commands }) => {
                  return commands.toggleWrap('blockquote');
              },
              unsetBlockquote: () => ({ commands }) => {
                  return commands.lift('blockquote');
              },
          };
      },
      addKeyboardShortcuts() {
          return {
              'Mod-Shift-b': () => this.editor.commands.toggleBlockquote(),
          };
      },
      addInputRules() {
          return [
              prosemirrorInputrules.wrappingInputRule(inputRegex, this.type),
          ];
      },
  });

  exports.Blockquote = Blockquote;
  exports.default = Blockquote;
  exports.inputRegex = inputRegex;

  Object.defineProperty(exports, '__esModule', { value: true });

})));
