// @tiptap/extension-blockquote/dist/tiptap-extension-blockquote.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@tiptap/core');
var prosemirrorInputrules = require('prosemirror-inputrules');

const inputRegex = /^\s*>\s$/gm;
const Blockquote = core.Node.create({
    name: 'blockquote',
    defaultOptions: {
        HTMLAttributes: {},
    },
    content: 'block*',
    group: 'block',
    defining: true,
    parseHTML() {
        return [
            { tag: 'blockquote' },
        ];
    },
    renderHTML({ HTMLAttributes }) {
        return ['blockquote', core.mergeAttributes(this.options.HTMLAttributes, HTMLAttributes), 0];
    },
    addCommands() {
        return {
            setBlockquote: () => ({ commands }) => {
                return commands.wrapIn('blockquote');
            },
            toggleBlockquote: () => ({ commands }) => {
                return commands.toggleWrap('blockquote');
            },
            unsetBlockquote: () => ({ commands }) => {
                return commands.lift('blockquote');
            },
        };
    },
    addKeyboardShortcuts() {
        return {
            'Mod-Shift-b': () => this.editor.commands.toggleBlockquote(),
        };
    },
    addInputRules() {
        return [
            prosemirrorInputrules.wrappingInputRule(inputRegex, this.type),
        ];
    },
});

exports.Blockquote = Blockquote;
exports.default = Blockquote;
exports.inputRegex = inputRegex;
